DataStructures/Graph/DirectedWeightedGraph.cs

```diff
@@ -0,0 +1,202 @@
+    public class DirectedWeightedGraph<T>
+    {
+        /// <summary>
+        ///     Default capacity of the graph.
+        /// </summary>
+        private const int DefaultCapacity = 10;
+
+        /// <summary>
+        ///     Capacity of the graph, indicates the maximum amount of vertices.
+        /// </summary>
+        private readonly int capacity;
+
+        /// <summary>
+        ///     Adjacency matrix which reflects the edges between vertices and their weight.
+        ///     Zero value indicates no edge between two vertices.
+        /// </summary>
+        private readonly double[,] adjacencyMatrix;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="DirectedWeightedGraph{T}"/> class.
+        /// </summary>
+        /// <param name="capacity">Capacity of the graph, indicates the maximum amount of vertices.</param>
+        public DirectedWeightedGraph(int capacity)
+        {
+            ThrowIfCapacityLessOne(capacity);
+
+            this.capacity = capacity;
+            Vertices = new List<Vertex<T>>();
+            adjacencyMatrix = new double[capacity, capacity];
+            Count = 0;
+        }
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="DirectedWeightedGraph{T}"/> class
+        ///     with default capacity equals 10.
+        /// </summary>
+        public DirectedWeightedGraph()
+        {
+            capacity = DefaultCapacity;
+            Vertices = new List<Vertex<T>>();
+            adjacencyMatrix = new double[capacity, capacity];
+            Count = 0;
+        }
+
+        /// <summary>
+        ///     Gets list of vertices of the graph.
+        /// </summary>
+        public List<Vertex<T>> Vertices { get; }
+
+        /// <summary>
+        ///     Gets current amount of vertices in the graph.
+        /// </summary>
+        public int Count { get; private set; }
+
+        /// <summary>
+        ///     Adds new vertex to the graph.
+        /// </summary>
+        /// <param name="data">Data of the vertex.</param>
+        /// <returns>Reference to created vertex.</returns>
+        public Vertex<T> AddVertex(T data)
+        {
+            ThrowIfOverflow();
+            var vertex = new Vertex<T>(data, Count, this);
+            Vertices.Add(vertex);
+            Count++;
+            return vertex;
+        }
+
+        /// <summary>
+        ///     Creates an edge between two vertices of the graph.
+        /// </summary>
+        /// <param name="startVertex">Vertex, edge starts at.</param>
+        /// <param name="endVertex">Vertex, edge ends at.</param>
+        /// <param name="weight">Double weight of an edge.</param>
+        public void AddEdge(Vertex<T> startVertex, Vertex<T> endVertex, double weight)
+        {
+            ThrowIfVertexNotInGraph(startVertex);
+            ThrowIfVertexNotInGraph(endVertex);
+
+            ThrowIfWeightZero(weight);
```

Comment:
I kept the possibility of negative weights of edges, so that Bellman-Ford algorithm (Shortest path algorithms including negative weights) can be implemented as well. So, should we keep negative weights or remove them?

Reply:
Let's keep them then